### ISSUE TITLE
Inject function parameters

### ISSUE THREAD

**Issue by StevenDeRoover** [introtocomputerscience/azure-function-autofac-dependency-injection#45]
Today, I started working with this Nuget package, and got the injection to work.

Previously, I used another DI solution, however, I needed to create a container manually in the Azure Function, and resolve "the first in line".
Fortunately, with this solution, this isn't needed anymore.

However: in my previous setup, I could give the ILogger (Microsoft.Extensions.Logging) - given as a parameter to the Azure Function - into a wrapper to my container.  The wrapper implemented my own "ILogger", as I didn't want to expose the Microsoft.Extensions library to other assemblies.

Is it possible to let the ILogger (even better all parameters of the Azure Function) being injected into a constructor?

**Comment by rawrspace** [comment]
That's an interesting question. Right now I pass the function name through to the constructor but I am not sure about the parameters. Let me have a look and I'll update this issue.

**Comment by StevenDeRoover** [comment]
I have found the solution: there's no need to update the AzureFunctions.Autofac package.

The solution: lazy loading using Autofac.  

All you need to do, is use a Func<params ..., TheClass> as parameter in the azure function.
Then, in the method of the function, use variableName(params ...).  Then, those params are used to construct the class.

**Comment by rawrspace** [comment]
Nice find. If you want to provide a code snippet I can add it to the documentation and credit you.

**Comment by StevenDeRoover** [comment]
Sure.

Let's say we have 2 services:
```
public interface IService    {    }

public interface IAnotherService    {    }
```
The first service requires the Logger of the Azure Function, and the second service:

```
public class Service : IService    {      
 public Service(IAnotherService anotherService, ILogger logger) {  }    
}
```
The Autofac bindings are the same as always:
```
DependencyInjection.Initialize(builder =>
            {
                builder.RegisterType<Service>().As<IService>();
                builder.RegisterType<AnotherService>().As<IAnotherService>();
            }, functionName);

```
In the Azure Function, we would want the IService injected as a Factory (Func<>) instead of directly injecting it:
```
[FunctionName("TestFunction")]
        public static async Task<HttpResponseMessage> Run([HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")]HttpRequestMessage req, ILogger log, [Inject]Func<ILogger, IService> serviceFactory)
        {
            var theService = serviceFactory(log);
            return req.CreateResponse(HttpStatusCode.OK, "");
        }
```
You get the instance of the service (including the ILogger) by calling the Func:
`var service = serviceFactory(log);`

Notice that, however we only provide the logger to the Func<>, the other dependencies are injected in the Service as well.  Another note: the order of the parameters in the Service is not important: the ILogger and IAnotherService can appear in any order.

Update 08/01/2020: 
this method has one side effect: the dependencies which you give through the factory (Func) will not be injected in further downwards dependencies (f.i.: the AnotherService would not get the ILogger).
There is a solution for that as well: Inject the Autofac's IComponentContext and register the dependencies.  Using this method, you don't even have to pass your dependencies through the Factory (even more: you better not :-)).

Inject the IComponentContext in the Azure Function:
`[Inject] IComponentContext context`

Then register the new dependency in Autofac:
`context.ComponentRegistry.Register(RegistrationBuilder.ForDelegate((c, p) => log).As<ILogger>().CreateRegistration());`

After that: call your factory (Func).  The ILogger will then be injected.

**Comment by rawrspace** [comment]
@StevenDeRoover I just merged in #48 which will provide the ability to more easily inject ILogger. Your method however is still interesting to me because of the factory aspect. Is that a native Autofac feature to provide you the factory when you provide the Func or did you register that Func somewhere?

**Comment by rawrspace** [comment]
This functionality is now available in [3.2.0](https://www.nuget.org/packages/AzureFunctions.Autofac/3.2.0)

**Comment by StevenDeRoover** [comment]
> @StevenDeRoover I just merged in #48 which will provide the ability to more easily inject ILogger. Your method however is still interesting to me because of the factory aspect. Is that a native Autofac feature to provide you the factory when you provide the Func or did you register that Func somewhere?

Yes, actually it is a basic usage of Autofac.  More on this on: https://autofaccn.readthedocs.io/en/latest/advanced/delegate-factories.html
